### DISCUSSION TITLE
Reduce white space between plots / add multiple plots to one frame card

### DISCUSSION THREAD

**Discussion by codeananda** [h2oai/wave#1359] [Q&A] [answered]
I want to add several plotly plots in a column but there is so much white space both in the frame cards themselves and between the cards. 

- How can I increase the amount of space the plots take up in each frame card?
- How can I reduce the amount of white space between plots? Best solution I can think of is to add multiple plots to one frame card. Is this possible?
- Lastly, are frame cards the right card for the job? Is there a slicker way to do it?

Minimal code example

```python
from h2o_wave import main, app, Q, ui
import numpy as np
import plotly.express as px
from plotly import io as pio

@app("/demo")
async def serve(q: Q):

    q.page["hist_1"] = ui.frame_card(box="1 1 4 3", title="", content=plot_hist(q))

    q.page["hist_2"] = ui.frame_card(box="1 4 4 3", title="", content=plot_hist(q))

    q.page["hist_3"] = ui.frame_card(box="1 7 4 3", title="", content=plot_hist(q))

    await q.page.save()

def plot_hist(q):
    fig = px.histogram(x=np.random.normal(size=1000))
    html = pio.to_html(fig, validate=False, include_plotlyjs="cdn")
    return html
```

Output
<img width="616" alt="Screenshot 2022-04-18 at 17 46 40" src="https://user-images.githubusercontent.com/51246969/163790983-5cc098c5-8ecf-4101-9f58-b512a868c03d.png">

**Comment by mturoci** [comment]
Thanks for the question @codeananda!

Please use our [discussions](https://github.com/h2oai/wave/discussions) for questions and issues for feature/bug reports.

TLDR: Use [native Wave plots](https://wave.h2o.ai/docs/widgets/plots/interval/#histogram) whenever possible, you will save yourself a lot of trouble.

> How can I increase the amount of space the plots take up in each frame card?
>How can I reduce the amount of white space between plots?

![image](https://user-images.githubusercontent.com/64769322/163948384-9e7ec7d0-38ce-4259-9822-6e9e6eb2db80.png)
You need to figure out in plotly. The output of plotly is HTML, that is later rendered in a `frame_card` which is just an `iframe` - element to display whatever HTML you provide. As you can see in the pic above, the card is filled to the full, but plotly for some reason decided to render the plot with fixed dimensions instead of responsive ones. Your options are:

* Figure out the right config in Plotly
* Reduce the height of card, but you will be left with small plots, which you probably do not desire.

>  Best solution I can think of is to add multiple plots to one frame card. Is this possible?

Yes, you can put as much HTML to plot_card as you wish.

> Lastly, are frame cards the right card for the job? Is there a slicker way to do it?

See, TLDR above.

**Comment by mtanco** [comment, marked as answer]
@codeananda Thanks for this question! And thanks to @mturoci on the plotly margin's direction. I have an app with some obnoxious white space that I have been meaning to deal with for months and this was good motivation. If you want to continue using Plotly and multiple plots, you can use the `fig.update_layout` function like:

![image](https://user-images.githubusercontent.com/4793561/164549522-5924d020-d990-44d6-b3ed-7caa34ada6b3.png)

```python
from h2o_wave import main, app, Q, ui
import numpy as np
import plotly.express as px
from plotly import io as pio

@app("/")
async def serve(q: Q):

    q.page["hist_1"] = ui.frame_card(box="1 1 4 3", title="Default Plotly Margins", content=plot_hist(change_margin=False))
    q.page["hist_2"] = ui.frame_card(box="1 4 4 3", title="No Plotly Margins", content=plot_hist())
    await q.page.save()


def plot_hist(change_margin=True):

    fig = px.histogram(x=np.random.normal(size=1000))
    if change_margin:
        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
    html = pio.to_html(fig, validate=False, include_plotlyjs="cdn")
    return html
```


If you do want to use plotly, and all plots in the same card, you can just "add" your html strings together. Please note in this example that I used the `height` parameter of Plotly Express, other wise the default is 100% which means the plot will grow to fill the card and your user will have to scroll to see each plot. 

![image](https://user-images.githubusercontent.com/4793561/164552783-e14fc65c-181f-4013-98f5-a7f6174be831.png)


```python
from h2o_wave import main, app, Q, ui
import numpy as np
import plotly.express as px
from plotly import io as pio


@app("/")
async def serve(q: Q):

    q.page["hist2"] = ui.frame_card(
        box="5 1 4 9",
        title="",
        content=plot_hist() + plot_hist()
    )
    await q.page.save()


def plot_hist() -> str:
    fig = px.histogram(x=np.random.normal(size=1000), height=300)
    fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
    html = pio.to_html(fig, validate=False, include_plotlyjs="cdn")
    return html
```

If you're interested in using H2O Wave plots, you can replicate your examples:

![image](https://user-images.githubusercontent.com/4793561/164555296-6d6826d3-2e2e-4abc-aca2-71643b2b1826.png)

```python
from h2o_wave import main, app, Q, ui, data
import numpy as np
import plotly.express as px
from plotly import io as pio


@app("/")
async def serve(q: Q):

    random_numbers = np.random.normal(size=1000)
    hist_data = np.histogram(random_numbers, bins=[-4, -3, -2, -1, 0, 1, 2, 3, 4])

    q.page["wave_plot"] = ui.plot_card(
        box="1 1 4 4",
        title="",
        data=data(
            fields=["counts", "values"],
            columns=[
                hist_data[0].astype('float'),
                hist_data[1].astype('float')
            ]),
        plot=ui.plot([ui.mark(type='interval', x='=values', y='=counts', x_title="x", y_title="count")])
    )

    q.page["plotly_plot"] = ui.frame_card(
        box="5 1 4 4",
        title="",
        content=plot_hist(random_numbers)
    )
    await q.page.save()


def plot_hist(plot_data) -> str:
    fig = px.histogram(x=plot_data, nbins=10)
    fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
    html = pio.to_html(fig, validate=False, include_plotlyjs="cdn")
    return html
```


And finally, if you want to use multiple wave plots in one card, you can use `ui.visualization` inside of a `ui.form_card`. 

Hope this helps!

**Reply by codeananda** [reply]
Wow! This is amazing!! Such a detailed and brilliantly helpful answer! Thank you @mtanco

**Reply by mtanco** [reply]
H2O Wave gives you lots of flexibility to do exactly what you want and integrate what you want, but it does take some getting started time :D happy to help
